File: typescript/safecracker.ts
import axios from 'axios';
import type { AxiosInstance } from 'axios';
import * as cheerio from 'cheerio';
import * as fs from 'fs';

const TARGET_URL = 'https://softwaretesting.nl';
const REPORT_FILE = './safecrack_report.md';
const DELAY_MS = 15;

interface TestResult {
    combo: string; 
    status: 'LEGITIMATE_CODE' | 'BUG_FOUND' | 'SECURELY_LOCKED';
    details: string; 
    latencyMs: number; 
    payloadBytes: number; 
    httpStatus: string;
}

async function createSession(id: number): Promise<[AxiosInstance, string]> {
    const instance = axios.create({ timeout: 5000 });
    const name = `Clean_Audit_Bot_T${id}`;

    const initGet = await instance.get(TARGET_URL);
    const sessionCookie = initGet.headers['set-cookie']?.map(c => c.split(';')).join('; ') || '';

    await instance.post(TARGET_URL, new URLSearchParams({
        'action': 'set_name', 'name': name
    }), { headers: { 'Content-Type': 'application/x-www-form-urlencoded', 'Cookie': sessionCookie } });

    return [instance, sessionCookie];
}

async function runCombo(instance: AxiosInstance, cookie: string, a: string, b: string, c: string, d: string): Promise<[string, string, number]> {
    const start = Date.now();
    const parameters = [['A', a], ['B', b], ['C', c], ['D', d]];

    for (const [param, value] of parameters) {
        if (DELAY_MS > 0) await new Promise(res => setTimeout(res, DELAY_MS));
        await instance.post(TARGET_URL, new URLSearchParams({
            'action': 'select', 'param': param, 'value': value
        }), { headers: { 'Content-Type': 'application/x-www-form-urlencoded', 'Cookie': cookie } });
    }

    if (DELAY_MS > 0) await new Promise(res => setTimeout(res, DELAY_MS));
    
    const res = await instance.post(TARGET_URL, new URLSearchParams({ 'action': 'add_attempt' }), {
        headers: { 'Content-Type': 'application/x-www-form-urlencoded', 'Cookie': cookie }
    });

    return [res.data, res.status.toString(), Date.now() - start];
}

async function main() {
    console.log("Generating combinatorial coverage dataset...");
    const pA = ['red', 'green', 'blue'];
    const pB = ['left', 'middle', 'right'];
    const pC = ['0', '1', '2'];
    const pD = ['alpha', 'beta', 'gamma'];

    const testCases: [string, string, string, string, string][] = [];
    const maxLen = Math.max(pA.length, pB.length, pC.length, pD.length);

    for (let i = 0; i < maxLen * maxLen; i++) {
        const a = pA[Math.floor(i / maxLen) % pA.length];
        const b = pB[i % pB.length];
        // Fixed: Swapped out 'pC.len' typo for valid 'pC.length' accessor property
        const c = pC[Math.floor(i / maxLen) % pC.length];
        const d = pD[(Math.floor(i / maxLen) + (i % maxLen)) % pD.length];
        testCases.push([a, b, c, d, 'Pairwise Matrix Rule']);
    }

    for (const a of pA) {
        for (const b of pB) {
            for (const c of pC) {
                for (const d of pD) {
                    const is3w = a !== 'red' && b === 'right' && d === 'alpha';
                    const is4w = a === 'red' && b === 'right' && c === '2' && d === 'gamma';
                    if (is3w || is4w) {
                        testCases.push([a, b, c, d, is3w ? 'Augmented: 3-Way Core Interaction' : 'Augmented: Strict 4-Way Glitch']);
                    }
                }
            }
        }
    }

    const logs: TestResult[] = [];
    let currentId = 1;
    console.log(`Running matrix sweep over ${testCases.length} isolated per-test sessions...`);

    for (const [a, b, c, d, label] of testCases) {
        try {
            const [instance, cookie] = await createSession(currentId);
            currentId++;

            const [body, status, latency] = await runCombo(instance, cookie, a, b, c, d);
            const combo = `${a} | ${b} | ${c} | ${d}`;
            const size = body.length;

            let evalStatus: 'LEGITIMATE_CODE' | 'BUG_FOUND' | 'SECURELY_LOCKED' = 'SECURELY_LOCKED';
            let diagnostics = 'Safe remained locked';

            const $ = cheerio.load(body);
            const txt = $('.display').text().toLowerCase();

            if (!txt.includes('closed')) {
                if (combo === 'red | left | 0 | alpha') {
                    evalStatus = 'LEGITIMATE_CODE';
                    diagnostics = 'Authorized standard route';
                } else {
                    evalStatus = 'BUG_FOUND';
                    diagnostics = label;
                }
            }

            logs.push({ combo, status: evalStatus, details: diagnostics, latencyMs: latency, payloadBytes: size, httpStatus: status });
        } catch (err) {
            console.error(`[ERROR] Execution broken at combo: ${a}-${b}-${c}-${d}`, err);
        }
    }

    let markdown = `# Production Safe Cracking Metrics & Comprehensive TypeScript Report\n\n`;
    markdown += `| Combination Profile (A, B, C, D) | Status | Diagnostics | Latency | Size | HTTP |\n`;
    markdown += `| :--- | :--- | :--- | :--- | :--- | :--- |\n`;
    for (const r of logs) {
        markdown += `| **${r.combo}** | \`${r.status}\` | ${r.details} | ${r.latencyMs}ms | ${r.payloadBytes} B | \`${r.httpStatus}\` |\n`;
    }

    fs.writeFileSync(REPORT_FILE, markdown);
    console.log(`Sweep complete! Verified results written cleanly to: ${REPORT_FILE}`);
}

main().catch(console.error);

File: rust/src/safecracker.rs
use std::collections::HashMap;
use std::env;
use std::fs::File;
use std::io::Write;
use reqwest::Client;
use scraper::{Html, Selector};
use tokio::time::{sleep, Duration};

const REPORT_FILE: &str = "./safecrack_report.md";
const MAX_ATTEMPTS_PER_SESSION: usize = 10; // Maximised to exactly 10 attempts per session block

struct TestResult {
    combo: String,
    status: String,
    details: String,
}

async fn create_authenticated_session(url: &str, batch_id: usize) -> Result<Client, Box<dyn std::error::Error>> {
    let client = Client::builder()
        .cookie_store(true)
        .build()?;

    let team_name = format!("Automated_Bot_Batch_{}", batch_id);
    println!("[SESSION] Spawning fresh incognito container context for: {}...", team_name);

    // Fixed: Force both inserts to use &str lifetimes to align types cleanly
    let mut startup_token = HashMap::new();
    startup_token.insert("action", "set_name");
    startup_token.insert("name", team_name.as_str());

    client.post(url).form(&startup_token).send().await?;
    Ok(client)
}

async fn submit_and_open_combo(client: &Client, url: &str, a: String, b: String, c: String, d: String, delay_ms: u64) -> Result<String, reqwest::Error> {
    let parameters = vec![("A", a), ("B", b), ("C", c), ("D", d)];

    for (param, value) in parameters {
        if delay_ms > 0 { sleep(Duration::from_millis(delay_ms)).await; }
        let mut form = HashMap::new();
        form.insert("action", "select".to_string());
        form.insert("param", param.to_string());
        form.insert("value", value);

        client.post(url).form(&form).send().await?;
    }

    if delay_ms > 0 { sleep(Duration::from_millis(delay_ms)).await; }
    let mut attempt_form = HashMap::new();
    attempt_form.insert("action", "add_attempt".to_string());

    let final_body = client.post(url)
        .form(&attempt_form)
        .send()
        .await?
        .text()
        .await?;

    Ok(final_body)
}

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let args: Vec<String> = env::args().collect();
    if args.len() < 2 {
        eprintln!("Usage: cargo run --bin safecracker <TARGET_URL> [--delay <ms>]");
        std::process::exit(1);
    }
    
    let target_url: &str = &args[1];
    
    let mut delay_ms: u64 = 0;
    if let Some(idx) = args.iter().position(|x| x == "--delay") {
        if idx + 1 < args.len() {
            delay_ms = args[idx + 1].parse().unwrap_or(0);
        }
    }

    println!("Connecting to target sequence engine: {}...", target_url);

    let p_a = vec!["red".to_string(), "green".to_string(), "blue".to_string()];
    let p_b = vec!["left".to_string(), "middle".to_string(), "right".to_string()];
    let p_c = vec!["0".to_string(), "1".to_string(), "2".to_string()];
    let p_d = vec!["alpha".to_string(), "beta".to_string(), "gamma".to_string()];

    let mut combinations_to_test = Vec::new();
    let max_len = p_a.len().max(p_b.len()).max(p_c.len()).max(p_d.len());
    
    // Suite 1: Pairwise configurations
    for i in 0..(max_len * max_len) {
        let a = p_a[(i / max_len) % p_a.len()].clone();
        let b = p_b[i % p_b.len()].clone();
        let c = p_c[(i / max_len) % p_c.len()].clone();
        let d = p_d[((i / max_len) + (i % max_len)) % p_d.len()].clone();
        combinations_to_test.push((a, b, c, d, "Pairwise Matrix Rule".to_string()));
    }

    // Suite 2: Discovered multi-way variants
    for a in &p_a {
        for b in &p_b {
            for c in &p_c {
                for d in &p_d {
                    let is_three_way = a != "red" && b == "right" && d == "alpha";
                    let is_four_way = a == "red" && b == "right" && c == "2" && d == "gamma";
                    
                    if is_three_way || is_four_way {
                        let label = if is_three_way { "Augmented: 3-Way Core Interaction" } else { "Augmented: Strict 4-Way Glitch" };
                        combinations_to_test.push((a.clone(), b.clone(), c.clone(), d.clone(), label.to_string()));
                    }
                }
            }
        }
    }

    let mut test_results: Vec<TestResult> = Vec::new();
    let mut current_batch = 1;
    let mut current_client = create_authenticated_session(target_url, current_batch).await?;
    let mut session_attempt_counter = 0;

    println!("Running matrix audit with auto-cookie session rotation...");

    for (a, b, c, d, rule_label) in combinations_to_test {
        let body = submit_and_open_combo(&current_client, target_url, a.clone(), b.clone(), c.clone(), d.clone(), delay_ms).await?;
        session_attempt_counter += 1;
        
        let combo = format!("{} | {} | {} | {}", a, b, c, d);

        let document = Html::parse_document(&body);
        let mut matched_status = String::new();

        if let Ok(selector) = Selector::parse(".display") {
            for element in document.select(&selector) {
                let inner_text = element.text().collect::<Vec<_>>().join(" ").to_lowercase();
                if !inner_text.contains("closed") {
                    if combo == "red | left | 0 | alpha" {
                        matched_status = "LEGITIMATE_CODE".to_string();
                    } else {
                        matched_status = "BUG_FOUND".to_string();
                    }
                }
            }
        }

        if matched_status == "LEGITIMATE_CODE" {
            test_results.push(TestResult { combo, status: "LEGITIMATE_CODE".to_string(), details: "Authorized standard route".to_string() });
        } else if matched_status == "BUG_FOUND" {
            test_results.push(TestResult { combo, status: "BUG_FOUND".to_string(), details: rule_label });
        }

        // Fixed: If we just performed the 10th attempt, clear cookies immediately before starting the next combination loop
        if session_attempt_counter >= MAX_ATTEMPTS_PER_SESSION {
            current_batch += 1;
            current_client = create_authenticated_session(target_url, current_batch).await?;
            session_attempt_counter = 0;
        }
    }

    let mut file = File::create(REPORT_FILE)?;
    writeln!(file, "# Production Safe Cracking Verification Matrix Report")?;
    writeln!(file, "\n| Combination Variant Profile (A, B, C, D) | Evaluation Status | Mechanism Diagnostics |")?;
    writeln!(file, "| :--- | :--- | :--- |")?;
    for res in test_results {
        writeln!(file, "| **{}** | `{}` | {} |", res.combo, res.status, res.details)?;
    }

    println!("Verification sweep complete. True defects cleanly stored in: {}", REPORT_FILE);
    Ok(())
}

